### PR TITLE
SALTO-1389, SALTO-1526: Safe deploy for Netsuite

### PR DESCRIPTION
**Pull request by nitzan-haim**
Add a change validator for Netsuite. This change validator catches changes made in the service on instances being deployed after an instance was fetched. (for further context refer to the ticket).
Added a flag `warnOnStaleWorkspaceData` to the deploy part of config (hidden from the config), to enable/disable the running of this validator. the default for this flag is false for now.

---

required refactor of NetSuite adapter's fetch in order to create the `fetchByQuery` method (used in new change validator).

---
_Release Notes_: 
Netsuite Adapter: Added a change validator to warn when attempting to change an element that was changed in the service since the last time the user fetched ("safe deploy").
---
_User Notifications_: 
NA

### CODE REVIEW

File: salto.code-workspace [hunk 1]
```diff
@@ -99,6 +99,9 @@
 		{
 			"name": "stripe-adapter",
 			"path": "packages/stripe-adapter"
+		},
+		{
+			"path": "temp"
```
Comment (nitzan-haim) [inline salto.code-workspace:104]: by mistake, will remove this from the PR

**Review by omrilit — COMMENTED** [review]
Generally looks great! I added some comments, most are nits but some are important.

Could you please make sure to update the documentation as well in `config_doc.md`?
Please pay attention that both the new flag and `deployReferencedElements` should be under the `deploy` section which is missing.

File: packages/netsuite-adapter/src/change_validators/safe_deploy.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  fetchByQuery?: FetchByQueryFunc) => {
+  const errors: ChangeError[] = []
+
+  if (fetchByQuery !== undefined) {
+    const modificationOrRemovalInstanceChanges = await awu(changes)
+      .filter(isRemovalOrModificationChange)
+      .filter(isInstanceChange)
+      .toArray()
+
+    const serviceInstances = await getMatchingServiceInstances(
```
[the remark this answers is not in the document]
Comment (nitzan-haim) [reply packages/netsuite-adapter/src/change_validators/safe_deploy.ts:64]: just opened a ticket exactly about this (SALTO-1526). Tomer said the chances of this use-case actually happening is small so we will keep it documented in the ticket in low priority and consider adding it in the future.

File: packages/netsuite-adapter/src/adapter.ts [hunk 3]
```diff
@@ -165,48 +168,28 @@ export default class NetsuiteAdapter implements AdapterOperations {
...
-    let fetchQuery = [
-      this.fetchInclude && buildNetsuiteQuery(this.fetchInclude),
-      this.fetchTarget && buildNetsuiteQuery(convertToQueryParams(this.fetchTarget)),
-      this.fetchExclude && notQuery(buildNetsuiteQuery(this.fetchExclude)),
-      this.skipList && notQuery(buildNetsuiteQuery(convertToQueryParams(this.skipList))),
-      notQuery(deprecatedSkipList),
-    ].filter(values.isDefined).reduce(andQuery)
-
-
+  public fetchByQuery = async (fetchQuery: NetsuiteQuery, progressReporter: ProgressReporter,
```
[the remark this answers is not in the document]
Comment (nitzan-haim) [reply packages/netsuite-adapter/src/adapter.ts:171]: you mean like this?
```
  public fetchByQuery: FetchByQueryFunc = async (
    fetchQuery: NetsuiteQuery,
    progressReporter: ProgressReporter,
    useChangesDetection: boolean
  ): Promise<FetchByQueryReturnType> => {
```

File: packages/netsuite-adapter/src/change_validator.ts [hunk 4]
```diff
@@ -48,14 +49,25 @@ const nonSuiteAppValidators: ChangeValidator[] = [
...
+  fetchByQuery?: FetchByQueryFunc
+  }) => ChangeValidator = ({ withSuiteApp, warnStaleData, fetchByQuery }) =>
+    async changes => {
+      const validators = warnStaleData
+        ? [...changeValidators, safeDeployValidator]
+        : [...changeValidators]
+      if (!withSuiteApp) validators.push(...nonSuiteAppValidators)
+
+      const changeErrors = _.flatten(await Promise.all(validators
+        .map(validator => validator(changes, fetchByQuery))))
```
[the remark this answers is not in the document]
Comment (nitzan-haim) [reply packages/netsuite-adapter/src/change_validator.ts:67]: talked in person and decided to keep the `ChangeValidator` and `QueryChangeValidator` separate, and run them seperately

File: packages/netsuite-adapter/config_doc.md [hunk 5]
```diff
@@ -40,10 +40,11 @@ netsuite {
...
-| deployReferencedElements                            | false                   | Deployment of a certain configuration element will include all elements referred by it
 | [client](#sdf-client-configuration-options)         | {} (no overrides)       | Configuration relating to the SDF client used to interact with netsuite
 | [suiteAppClient](#salto-suiteapp-client-configuration-options)             | {} (no overrides)       | Configuration relating to the Salto SuiteApp client used to interact with netsuite
 | [deploy](#salto-deploy-flags)                       | undefined. set all deploy's flags to their default value        | Configuration deploy optional flags
+| deploy.deployReferencedElements                            | false                   | Deployment of a certain configuration element will include all elements referred by it
+| deploy.warnOnStaleWorkspaceData                            | false                   | If assigned 'true' runs a validation upon deploy which warns the user if the changes override other changes made in the service since the last fetch
```
Comment (nitzan-haim) [inline packages/netsuite-adapter/config_doc.md:47]: @omrilit  please verify that this is the correct way to add the flags?

File: packages/netsuite-adapter/src/change_validators/safe_deploy.ts [hunk 6]
```diff
@@ -28,61 +32,118 @@ export type FetchByQueryReturnType = {
...
+  baseInstances: InstanceElement[],
+  fetchByQuery: FetchByQueryFunc
+): Promise<Record<string, InstanceElement>> => {
+  // TODO: we currently support only SDF types (i.e custom types and file cabinet),
+  // and not yet suiteapp types (SALTO-1531)
+
+  const filePaths = baseInstances
+    .filter(isFileCabinetInstance)
+    .filter(inst => inst.value[PATH] !== undefined)
+    .map(inst => inst.value[PATH])
```
Comment (nitzan-haim) [inline packages/netsuite-adapter/src/change_validators/safe_deploy.ts:64]: @omrilit please make sure I got the identifiers right (`PATH` and `SCRIPT_ID` for file cabinet types and custom types respectively)

**Review by omrilit — APPROVED** [review]

File: packages/netsuite-adapter/config_doc.md [hunk 7]
```diff
@@ -40,10 +40,11 @@ netsuite {
 | ----------------------------------------------------| ------------------------| -----------
 | [fetch.include](#fetch-include-configuration-options)                                  | Include everything                               | Specifies items to fetch. Items that do not match any of the include criteria will not be fetched
```
[the remark this answers is not in the document]
Comment (nitzan-haim) [reply packages/netsuite-adapter/config_doc.md:41]: it's broken (the `fetch.exclude` reference is also broken)

[the remark this answers is not in the document]
Comment (nitzan-haim) [reply packages/netsuite-adapter/config_doc.md:41]: opened SALTO-1539 because I wasn't sure if we want to delete them or have them lead to config options.

**Pull request merged**